cluster-logging.md

```diff
@@ -0,0 +1,110 @@
+# Cluster Logging and OpenTelemetry
+
+This is the protocol and semantic conventions documentation for Red Hat OpenShift Logging.
+
+## Forwarding and Ingestion Protocol
+
+Red Hat OpenShift Logging provides a log collection and forwarding solution that is capable of writing logs to OpenTelemetry endpoints using OTLP. [OTLP](https://opentelemetry.io/docs/specs/otlp/) is the *protocol* for encoding, transporting, and delivering telemetry data.  This product additionally is capable of providing a Loki storage deployment that provides an OTLP endpont to ingest log streams.  This document defines the semantic conventions associated with the logs collected from the various sources of an OpenShift cluster.
+
+**Note:** Logs are forwarded using OTLP/HTTP as defined by the OpenTelemetry Observability Framework.  It uses Protobuf payloads encoded in JSON format.
+
+
+## Semantic Conventions
+
+The log collector provided by this solution collects the following log streams:
+
+* Container logs
+* Cluster node journal logs
+* Cluster node auditd logs 
+* Kubernetes and OpenShift API server logs
+* OpenShift Virtual Network (OVN) logs
+
+These streams are forwarded using the sementatic conventions defined by [OpenTelemetry semantic attributes](https://github.com/open-telemetry/semantic-conventions/tree/main/docs). The semantic conventions in OpenTelemetry define a *Resource* as an immutable representation of the entity producing telemetry as *Attributes*. For example, a process producing telemetry that is running in a container has a container_name, a cluster_id, a pod_name, a namespace, and possibly a deployment or app_name. All of these *Attributes* are included in the *Resource* object.  This grouping and reducing of common attributes is a powerful tool when sending logs as telemetry data.
+
+
+The following sections define the attributes that are forwarded.
+
+
+### Common Attributes
+
+All log streams include a minimal set of attributes.
+
+Resource Attributes:
+```
+cluster.id
+node.name
+openshift.log.source
+openshift.log.type
+```
+
+Log Attributes:
+```
+body
+timeUnixNano
+severityNumber
+observedTimeUnixNano
+```
+
+### Kubernetes Container Logs
+
+Resource Attributes:
+```
+k8s.cluster.uid
+k8s.container.name
+k8s.namespace.name
+k8s.pod.name
+```
```

Comment:
What about the following? Do we want to start adding them? They make writing LogQL label selectors that address all the pods of an app easier (See comparison below).
```
k8s.deployment.name
k8s.statefulset.name
k8s.daemonset.name
k8s.cronjob.name
k8s.replicaset.name
k8s.job.name
```

Comparison:
```
{k8s_deployment_app="the-app"} | ...

vs

{k8s_pod_name=~"the-app-.*"}
```

Reply:
> k8s.deployment.name
k8s.statefulset.name
k8s.daemonset.name
k8s.cronjob.name
k8s.replicaset.name
k8s.job.name

I only left these out because the Loki config had a comment regarding the OTEL collector adding them.  I wasn't exactly certain how to express these would be conditional and would be absent for vector collectors.  I welcome suggestions

Reply:
I think we can add these in vector also, the values are all in the the Pod's OwnerRef. Need to look at vector's k8s adapter to see if it will do this.
If we do this then Loki's default service.name logic will use the owner name as the service name, which solves the service.name problem.

Reply:
Added, but not that unless these are discoverable by a pod annotation or other meta that comes in with the pod, vector will not be capable of adding them unless we modify the vector code.